### PR TITLE
Feature/future compat

### PR DESCRIPTION
**Pull request by axel22**
Mostly tests for the scala.concurrent package, and a couple of bugfixes.

### CODE REVIEW

File: src/library/scala/concurrent/ExecutionContext.scala [hunk 1]
```diff
@@ -47,11 +47,18 @@ object ExecutionContext {
...
+  def fromExecutorService(e: ExecutorService, reporter: Throwable => Unit = defaultReporter): ExecutionContext with Executor = new impl.ExecutionContextImpl(e, reporter)
   
   /** Creates an `ExecutionContext` from the given `Executor`.
    */
-  def fromExecutor(e: Executor): ExecutionContext with Executor = new impl.ExecutionContextImpl(e)
+  def fromExecutor(e: Executor, reporter: Throwable => Unit = defaultReporter): ExecutionContext with Executor = new impl.ExecutionContextImpl(e, reporter)
+  
+  def defaultReporter: Throwable => Unit = {
+    // `Error`s are currently wrapped by `resolver`.
+    // Also, re-throwing `Error`s here causes an exception handling test to fail.
```
Comment (viktorklang) [inline src/library/scala/concurrent/ExecutionContext.scala:58]: Which test?

File: src/library/scala/concurrent/Future.scala [hunk 2]
```diff
@@ -384,18 +384,18 @@ trait Future[+T] extends Awaitable[T] {
...
-
-    that onFailure {
-      case f => p failure f
+      case Right(r) =>
+        that onSuccess {
+          case r2 => p success ((r, r2))
+        }
+        that onFailure {
+          case f => p failure f
+        }
```
Comment (viktorklang) [inline src/library/scala/concurrent/Future.scala:396]: Use onComplete instead

File: src/library/scala/concurrent/impl/AbstractPromise.java [hunk 3]
```diff
@@ -9,13 +9,32 @@
...
+    
+    final static long _refoffset;
+    
+    static {
+	try {
+	    _refoffset = Unsafe.instance.objectFieldOffset(AbstractPromise.class.getDeclaredField("_ref"));
+	} catch (Throwable t) {
+	    throw new ExceptionInInitializerError(t);
+	}
+    }
```
Comment (viktorklang) [inline src/library/scala/concurrent/impl/AbstractPromise.java:28]: formatting

File: src/library/scala/concurrent/ExecutionContext.scala (hunk 1 above)
Comment (axel22) [inline src/library/scala/concurrent/ExecutionContext.scala:58]: ## I think I didn't write that comment, just copied it.

Aleksandar Prokopec
LAMP, IC, EPFL

Sent from my Android phone with K-9 Mail. Please excuse my brevity.

viktorklang reply@reply.github.com wrote:

> /*\* Creates an `ExecutionContext` from the given `Executor`.
> */
> - def fromExecutor(e: Executor): ExecutionContext with Executor = new impl.ExecutionContextImpl(e)
> - def fromExecutor(e: Executor, reporter: Throwable => Unit = defaultReporter): ExecutionContext with Executor = new impl.ExecutionContextImpl(e, reporter)
> - 
> - def defaultReporter: Throwable => Unit = {
> - // `Error`s are currently wrapped by `resolver`.
> - // Also, re-throwing `Error`s here causes an exception handling test to fail.

Which test?

---

Reply to this email directly or view it on GitHub:
https://github.com/scala/scala/pull/454/files#r758713

File: src/library/scala/concurrent/impl/AbstractPromise.java [hunk 4]
```diff
@@ -9,13 +9,32 @@
...
+	return Unsafe.instance.compareAndSwapObject(this, _refoffset, oldState, newState);
+    }
+    
+    protected final Object getState() {
+	return _ref;
+    }
+    
     protected final static AtomicReferenceFieldUpdater<AbstractPromise, Object> updater =
-            AtomicReferenceFieldUpdater.newUpdater(AbstractPromise.class, Object.class, "_ref");
+	AtomicReferenceFieldUpdater.newUpdater(AbstractPromise.class, Object.class, "_ref");
```
Comment (viktorklang) [inline src/library/scala/concurrent/impl/AbstractPromise.java:39]: This should be deleted

File: src/library/scala/concurrent/impl/Promise.scala [hunk 5]
```diff
@@ -77,8 +76,8 @@ object Promise {
   /** Default promise implementation.
    */
   class DefaultPromise[T](implicit val executor: ExecutionContext) extends AbstractPromise with Promise[T] { self =>
-    updater.set(this, Nil) // Start at "No callbacks" //FIXME switch to Unsafe instead of ARFU
-
+    updateState(null, Nil) // Start at "No callbacks" //FIXME switch to Unsafe instead of ARFU
```
Comment (viktorklang) [inline src/library/scala/concurrent/impl/Promise.scala:79]: Remove FIXME as it doesn't make sense.

File: src/library/scala/concurrent/impl/Promise.scala [hunk 6]
```diff
@@ -99,7 +98,7 @@ object Promise {
           isCompleted
       }
       //FIXME do not do this if there'll be no waiting
```
Comment (viktorklang) [inline src/library/scala/concurrent/impl/Promise.scala:100]: This FIXME isn't supposed to be at this line anymore

File: src/library/scala/concurrent/util/Unsafe.java [hunk 7]
```diff
@@ -6,22 +6,25 @@
 **                          |/                                          **
 \*                                                                      */
 
+package scala.concurrent.util;
```
Comment (viktorklang) [inline src/library/scala/concurrent/util/Unsafe.java:9]: IS this public API now?

File: src/library/scala/concurrent/ExecutionContext.scala (hunk 1 above)
Comment (viktorklang) [inline src/library/scala/concurrent/ExecutionContext.scala:58]: Can the offender explain him/herself? ;-)

**Comment by adriaanm** [comment]
@axel22, @viktorklang I'm not sure this is the only issue that keeps us from running on google app engine, but the change from `AtomicReferenceFieldUpdater` to `compareAndSwapObject` broke GAE (https://issues.scala-lang.org/browse/SI-8362). Could you explain the motivation (I mean, I assume it's performance, but the meta-point is that it would be good to have that be explicit in the code/docs/commit message, so I can answer bug reports on your behalf ;-))?

**Comment by viktorklang** [comment]
We use Unsafe for performance and compatibility reasons (performance in the sense that the FieldUpdaters are considerably slower due to their design, this has been frequently discussed on concurrency-interest) and compatibility in the sense that our embedded version of ForkJoinPool is written using Unsafe and switching to FieldUpdaters would not only make updates to the FJP code riskier (the FJP code is extremely hard to maintain, have a look) as well as make it slower, which is detrimental for an execution engine).

If someone has a potential solution with a back-off strategy that tries to use Unsafe but falls back to FieldUpdaters if there is any issues, that doesn't carry any runtime performance penalty, I'm all ears!

Also, since Akka also heavily uses Unsafe it wouldn't work on GAE either.

In any case, have we stated GAE compliance? (In which case we should run automated tests on GAE)
According to GAE, Scala works: https://developers.google.com/appengine/kb/general#language

**Comment by adriaanm** [comment]
We don't promise GAE support, but it would be nice if we could if the cost is low.
My Java is a bit rusty, but how about:

```
    private final static long _refoffset;
    private final static boolean _useUnsafe;

    static {
      boolean ok = true;
      long ofs = 0;
      try {
        ofs = Unsafe.instance.objectFieldOffset(AbstractPromise.class.getDeclaredField("_ref"));
      } catch (Throwable t) {
        // fallback to AtomicReferenceFieldUpdater if Unsafe is not available, e.g. on GAE (SI)
        ok = false;
      }
      _useUnsafe = ok; // TODO: check _refoffset != sun.misc.Unsafe.INVALID_FIELD_OFFSET
      _refoffset = ofs;
    }

    protected final boolean updateState(Object oldState, Object newState) {
      if (_useUnsafe) return Unsafe.instance.compareAndSwapObject(this, _refoffset, oldState, newState);
      else            return updater.compareAndSet(this, oldState, newState);
    }
```

**Comment by adriaanm** [comment]
I assume the JIT will quickly get rid of the if on a final static boolean(?).

**Comment by viktorklang** [comment]
Could work, but it's a lot of ceremony and is error prone (X fields and Y different uses (CAS, getAndSet etc)). Can we work out a solution where we can get to:

private final static IntUpdater _ref = Updater.create(AbstractPromise.class, "_ref");

And make sure that things will get inlined and intrinsified properly, then we're all good.
We'll need to run tests on GAE tho.

**Comment by viktorklang** [comment]
Would make a ton of sense to create something like this, private to the Scala StdLib.
That would open up the possibility of another idea, to provide language-level atomics in Scala, eliminating the need for Unsafe/Updaters in user code.

something like:

```
@atomic var _ref: AnyRef = null
_ref.getAndSet("foo")
_ref.compareAndSet("foo", "bar")
```

**Comment by adriaanm** [comment]
I agree that would be ideal & really cool. I guess I got a little carried away, let's schedule considering a minimal fix for 2.11.1, as it seems it would be binary compatible anyway.

**Comment by adriaanm** [comment]
I'm kind of intrigued by this, so I read [the thread you alluded to](http://cs.oswego.edu/pipermail/concurrency-interest/2011-November/008412.html) and tried to encapsulate the fallback to the updater in another class. 

It seems that the whole point, though, is better performance, so we don't want to add any indirection. Thus, it seems unwise to encapsulate the CAS behaviour in a new `Updater` class. Then we might as well use `AtomicReferenceFieldUpdater<U,W>`. 

By the way, `AbstractPromise` is a bit sloppy: it still has the dead `_updater` field, so I'm a little less inclined to favor the status quo.

Alternative updater below because I already wrote all this stuff down anyway and I'd feel less guilty about spending time on this if at least there's a record of some sort :-)

``` scala
object FastAtomicUpdater {
  def apply[T, V](tclass: Class[T], vclass: Class[V], fieldName: String): AtomicReferenceFieldUpdater[T, V] =
    try new AtomicUpdaterUnsafe(Unsafe.instance.objectFieldOffset(tclass.getDeclaredField(fieldName)))
    catch { case _ : Throwable => // ok ok, should be NonFatal
      // fallback to AtomicReferenceFieldUpdater if Unsafe is not available, e.g. on GAE (SI)
      AtomicReferenceFieldUpdater.newUpdater(tclass, vclass, fieldName)
    }
}


final class AtomicUpdaterUnsafe[T, V](fieldOffset: Long) extends AtomicReferenceFieldUpdater[T, V] {
  def compareAndSet(target: AnyRef, oldState: AnyRef, newState: AnyRef) =
    Unsafe.instance.compareAndSwapObject(target, fieldOffset, oldState, newState)

  // ...
}
```

**Comment by carey** [comment]
(I reported the issue on the Scala Jira.)

I have Scala code running perfectly well on Google App Engine, without any threading for now. Even single-threaded we're not exceeding the App Engine request deadlines at this point, and if I did want to use multiple threads I could use the `ExecutorService` directly.

`AtomicReferenceFieldUpdater` is not final, so `AtomicUpdaterUnsafe` could extend it while keeping the method short enough for it to be inlined, and substituting the official implementation otherwise. I did some testing with an implementation like that below, and the `lock cmpxchg` was inlined into the caller:

```
@Override
public boolean compareAndSet(T obj, Object expect, Object update) {
    return scala.concurrent.util.Unsafe.instance.compareAndSwapObject(obj, offset, expect, update);
}
```

I expect this could be written in Scala, but I didn't want to complicate things.

**Comment by viktorklang** [comment]
Whatever we do, we need to make sure we use Unsafe if it's available; and that the overhead will be that of what it currently is, i.e. it gets inlined and intrinsified properly)
We also need to check if we need to backport any fixes to 2.10.4(?) (or less likely, 2.9.5)

Thanks for being on top of this Adriaan

**Comment by adriaanm** [comment]
@carey, thanks for investigating! Any chance you could write this up as a PR or detail your investigation? How do you verify the cmpxchg gets emitted & inlined? (I'm more of a front-end guy.)

I think it would be nice (literally, "nice", since we don't really officially promise this) not to regress on our GAE support.

**Comment by carey** [comment]
Hotspot has an option to disassemble and dump the assembly code that it generates; in this case I used `-XX:+UnlockDiagnosticVMOptions -XX:CompileCommand=print,*TestUpdates.update` where the `update` method is where I hoped to get the atomic operation inlined. Some of the output looks like this:

```
  0x0245e03a: lock cmpxchg %esi,(%edi)
  0x0245e03e: mov    $0x1,%esi
  0x0245e043: je     0x0245e04e
  0x0245e049: mov    $0x0,%esi
```

leaving out how the `getfield` and `getstatic` byte code was implemented to get hold of the Unsafe instance and the field offset. There's other stuff there that I don't understand, hopefully just checking for null pointers and object types.

I can write up some more details of my testing later.

**Comment by som-snytt** [comment]
@adriaanm front-end guy -> "front man."  Also, "nice" literally:
http://etymonline.com/index.php?term=nice

**Comment by adriaanm** [comment]
Interesting evolution for such a small word. I guess I'd call myself a nice (renaissance) man then. Also, innocent in this particular pr.

**Comment by axel22** [comment]
Agree with everything said - if we could have our own `FieldUpdater`s available in the stdlib that fall back to `ARFU` and make sure they are properly inlined, I see a few other places we could use them, e.g.: https://github.com/scala/scala/blob/v2.10.3/src/library/scala/collection/concurrent/INodeBase.java#L19

**Comment by viktorklang** [comment]
Sounds great, I'm happy to review any PR that tries to address this.

**Comment by carey** [comment]
My testing was slowed down by finding that the Java 6 Hotspot compiler is quite a bit worse than Java 7, and can't see past abstract classes or checking against null.

The best option is maybe the simplest; if `AbstractPromise._refoffset` is set to -1 when `Unsafe` isn't available, then it seems that Hotspot will compile the only possible branch of `if (_refoffset == -1)`.

**Comment by viktorklang** [comment]
Manually fiddling with offsets means duplicating this code whenever/wherever (we're meant to be together — Shakira) Unsafe is used. I'm much more in favor of subclassing the Updaters if we can get it to inline and intrinsify properly.

**Comment by carey** [comment]
The Java 6 _Client_ VM doesn't inline alternate subclasses, as at https://github.com/carey/hotspot-tests/blob/444cd0c64d030cbfa8b78fdc6a5b5f967e99e53e/hotspot-tests/src/nz/geek/carey/unsafe/abstractclasses/BaseClass.java. Only 32-bit Java on Windows uses the client VM, where threading performance of Scala code is probably not a major concern, especially when it only seems to be Java 6.

See https://gist.github.com/carey/9444544 for the compiler output, which I saved this time.

**Comment by viktorklang** [comment]
I agree with the comment on client vm for Java 6. How does it behave with Tiered Compilation?

**Comment by carey** [comment]
I haven't been able to see any difference in my tests between the client VM and tiered compilation in Java 6 and Java 7, and Java 8 says:

```
Java HotSpot(TM) Client VM warning: TieredCompilation is disabled in this release.
```

**Comment by ijuma** [comment]
You have to use -server to use TieredCompilation. There is a tiered policy that determines which compiler gets used when, see http://hg.openjdk.java.net/jdk7/jdk7/hotspot/file/5d8f5a6dced7/src/share/vm/runtime/advancedThresholdPolicy.cpp for an example.

**Comment by viktorklang** [comment]
I am experimenting with this for SIP-23:

https://github.com/viktorklang/scala/pull/6

Sadly, JSR166 relies on things that are not exposed on AtomicXFieldUpdaters (like relaxed writes), which means that there be uglies.

Also, I see a huge risk in modifying JSR166 classes (as well as the investment into testing and verifying UnsafeAtomicXUpdaters).

**Comment by viktorklang** [comment]
@adriaanm @carey I had a stab at implementing this but sadly I'll have to abort the effort: I just don't see how it will be possible to use AtomicXFieldUpdaters to do offset-based (relaxed!) writes into arrays, as done in the ForkJoinPool.java[1](https://github.com/scala/scala/blob/master/src/forkjoin/scala/concurrent/forkjoin/ForkJoinPool.java#L1053). See my attempt here[2].

[2]:
https://github.com/viktorklang/scala/pull/6/files#diff-9515f7f1449cd847c90a203ee46840ccR20

**Comment by adriaanm** [comment]
Thanks for the update, @viktorklang.

**Pull request merged**